sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -1561,6 +1563,71 @@ case class GpuFlattenArray(child: Expression) extends GpuUnaryExpression with Nu
   }
 }
 
+case class GpuArrayDistinct(child: Expression) extends GpuUnaryExpression with NullIntolerantShim {
+  private def childDataType: ArrayType = child.dataType.asInstanceOf[ArrayType]
+  override def nullable: Boolean = child.nullable || childDataType.containsNull
+  override def dataType: DataType = child.dataType
+
+  private def normalizeNaNs(input: ColumnView, elementType: DataType): ColumnVector = {
+    withResource(input.isNan()) { isNaN =>
+      val canonicalNaN = if (elementType == FloatType) {
+        Scalar.fromFloat(Float.NaN)
+      } else {
+        Scalar.fromDouble(Double.NaN)
+      }
+      withResource(canonicalNaN) { canonicalNaN =>
+        isNaN.ifElse(canonicalNaN, input)
+      }
+    }
+  }
+
+  private def floatingPointArrayDistinct(
+      input: GpuColumnVector,
+      elementType: DataType): ColumnVector = {
+    val intermediateDType = if (elementType == FloatType) { DType.INT32 } else { DType.INT64 }
+    val originalDType = if (elementType == FloatType) { DType.FLOAT32 } else { DType.FLOAT64 }
+    val arrayElements = input.getBase.getChildColumnView(0)
+    
+    withResource(arrayElements) { _ =>
+      val normalizedElements = normalizeNaNs(arrayElements, elementType)
+      withResource(normalizedElements) { _ =>
```

Comment:
Try to break up the nesting for readability and most of all to free resources once no longer needed.